In [1]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

In [2]:
breast_cancer = load_breast_cancer()
breast_cancer_data = breast_cancer.data        # 학습지
breast_cancer_label = breast_cancer.target     # 정답지

In [4]:
breast_cancer.target_names    # 악성, 양성

array(['malignant', 'benign'], dtype='<U9')

In [5]:
import pandas as pd
breast_cancer_df = pd.DataFrame(data = breast_cancer_data, columns = breast_cancer.feature_names)
breast_cancer_df.describe()

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension  ...  worst radius  \
count     569.000000              569.000000  ...    569.000000   
mean        0.181162                0.062798  ...     16.269190   
std         0.027414                0.007060  ...      4.833242   
min         0.106000                0.049960  ...      7.930000   
25%         0.161900                0.057700  ...     13.010000   
50%         0.179200                0.061540  ...     14.970000   
75%         0.195700                0.066120  ...     18.790000   
max         0.304000                0.097440  ...     36.040000   

       worst texture  worst perimeter   worst area  worst smoothness  \
count     569.000000       569.000000   569.000000        569.000000   
mean       25.677223       107.261213   880.583128          0.132369   
std         6.146258        33.602542   569.356993          0.022832   
min        12.020000        50.410000   185.200000          0.071170   
25%        21.080000        84.110000   515.300000          0.116600   
50%        25.410000        97.660000   686.500000          0.131300   
75%        29.720000       125.400000  1084.000000          0.146000   
max        49.540000       251.200000  4254.000000          0.222600   

       worst compactness  worst concavity  worst concave points  \
count         569.000000       569.000000            569.000000   
mean            0.254265         0.272188              0.114606   
std             0.157336         0.208624              0.065732   
min             0.027290         0.000000              0.000000   
25%             0.147200         0.114500              0.064930   
50%             0.211900         0.226700              0.099930   
75%             0.339100         0.382900              0.161400   
max             1.058000         1.252000              0.291000   

       worst symmetry  worst fractal dimension  
count      569.000000               569.000000  
mean         0.290076                 0.083946  
std          0.061867                 0.018061  
min          0.156500                 0.055040  
25%          0.250400                 0.071460  
50%          0.282200                 0.080040  
75%          0.317900                 0.092080  
max          0.663800                 0.207500  

[8 rows x 30 columns]

In [6]:
# 학습시키기
X_train, X_test, y_train, y_test = train_test_split(breast_cancer_data, breast_cancer_label, test_size=0.2, random_state=36)

In [8]:
# 여러 가지 모델 평가하기

### 1) Decision Tree
from sklearn.tree import DecisionTreeClassifier
decision_tree = DecisionTreeClassifier(random_state=80)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print("* Decision Tree 모델")
print(classification_report(y_test, y_pred))
print("\n")


### 2) Random Forest
from sklearn.ensemble import RandomForestClassifier
random_forest = RandomForestClassifier(random_state=80)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print("* Random Forest 모델")
print(classification_report(y_test, y_pred))
print("\n")


### 3) SVM
from sklearn import svm
svm_model = svm.SVC(random_state=80)
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print("* SVM 모델")
print(classification_report(y_test, y_pred))
print("\n")


### 4) SGD Classifier
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier(random_state=80)
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print("* SGD Classifier 모델")
print(classification_report(y_test, y_pred))
print("\n")


### 5) Logistic Regression
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression(random_state=80)
logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print("* Logistic Regression 모델")
print(classification_report(y_test, y_pred))

* Decision Tree 모델
              precision    recall  f1-score   support

           0       0.91      1.00      0.95        39
           1       1.00      0.95      0.97        75

    accuracy                           0.96       114
   macro avg       0.95      0.97      0.96       114
weighted avg       0.97      0.96      0.97       114



* Random Forest 모델
              precision    recall  f1-score   support

           0       0.95      0.95      0.95        39
           1       0.97      0.97      0.97        75

    accuracy                           0.96       114
   macro avg       0.96      0.96      0.96       114
weighted avg       0.96      0.96      0.96       114



* SVM 모델
              precision    recall  f1-score   support

           0       1.00      0.79      0.89        39
           1       0.90      1.00      0.95        75

    accuracy                           0.93       114
   macro avg       0.95      0.90      0.92       114
weighted avg       0.94

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


유방암 진단의 경우 악성인 환자를 놓치지 않는 것이 중요하기 때문에 양성을 음성으로 잘못 판단하면 안된다.  
따라서 이 경우에는 재현율이 매우 중요하다.  

첫 번째 모델인 Decision Tree의 경우 인덱스 0, 즉 악성(malignant)에 대한 재현율이 100%를 기록하였다.  
따라서 유방암 진단에 적합한 모델이라고 할 수 있다.  

두 번째 모델인 Random Forest는 첫 번째 모델과 정확도(Accuracy)가 96%로 같으나,  
재현율을 보았을 때 악성에 대한 재현율이 95%로 살짝 떨어져서 적합하지 못하다.

그 외 나머지 모델은 전체적으로 위 모델보다 기록이 낮으므로 선택하지 않는게 좋다.